rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+The key `/migrations/<migration_number>` will store the history of a migration.
+
+Cluster admins may view the migration history via `tctl migrations ls`.
+
+### Migrating Keys
+
+All migrations MUST not be performed in place, instead migrations MUST migrate both the keys and values in the backend.
+For example, to migrate the data stored in key `some/path/to/resource/<ID>` we must leave the original value at `some/path/to/resource/<ID>` as is and write the migrated value to `some/new/path/to/resource/<ID>`.
+This allows older versions of Teleport to still operate on `some/path/to/resource/<ID>`, while newer versions can first attempt reads from `some/new/path/to/resource/<ID>`, and fall back to reading from `some/path/to/resource/<ID>` if the migrated key does not exist yet.
+
+Migrations MUST also not delete the key being migrated for at least one major version of Teleport. This allows users to upgrade to and downgrade from a single major version, but prevents any downgrading beyond that.
+Continuing from the example above, if `some/path/to/resource/<ID>` is migrated to `some/new/path/to/resource/<ID>` in v1.0.0 a follow up migration should be added in v2.0.0 to delete the keys under `some/path/to/resource` that were migrated in v1.0.0.
+
+Keys should be versioned to determine which version of a resource exists at any given key range. A key prefix of `/type/version/subkind/name` should be used where possible to have uniformity.
+For example if nodes were migrated from `/nodes/default/<UUID>` it should be to `/nodes/v2/default/<UUID>`.
+
+### Backward Compatibility
+
+Migrations have historically been backward incompatible operations. Migrations altered the data in place without changing the key, which can prevent any versions prior to the migration from being able to unmarshal the value into the resource representation. The only way to downgrade in this scenario was to restore the backend from a backup prior to the migration, attempt to manually rollback the migration, or deleting the entire key range that was migrated. Moving forward it will be possible to downgrade a single major version without being impacted by a migration.
+
+### Testing Migrations
+
+While the framework laid out in this RFD allows migrations to be applied in a deterministic manner, it does not provide a uniform rule or process for any code that is impacted by a migration. To ensure that a migration is functional testing should consider a wide range of simultaneous versions in a cluster in accordance to our version compatibility matrix. Imagine that we are going to introduce a migration in v3.0.0, we must test the following for an extended period of time(10m) to ensure all supported versions are functional:
+
+| Auth 1 | Auth 2  | Proxy   | Agents  |
+| ------ | ------- | ------- | ------- |
+| v3.0.0 | v3.0.0  | v3.0.0  | v3.0.0  |
+| v3.0.0 | <v3.0.0 | <v3.0.0 | <v3.0.0 |
```

Comment:
I think this will create a problem:

* The auth 1 (new) writes the resource with the new field
* The auth 2 (old) reads the resource skipping the field, returns to the client, client sends update
* The auth 2 (old) will write the resource deleting the field 
* The auth 1 (new) will read the resource loosing the value

To avoid this problem, we can create a system that will enforce the following invariant - only one auth server version is writing and reading during upgrade. Here is how we can do it:

* Time 1. Auth servers version v0 are running
* Time 2. Auth servers version v0 and v1 are running, traffic is only sent to v0
* Time 3. Auth servers version v0 and v1 are running, traffic is only sent to v1
* Time 4. Auth servers version v1 are running, traffic is only sent to v1.

If we assume this invariant, migrations become simpler. If we assume we will use the same key, but[ attribute value will be versioned](https://www.dynamodbguide.com/anatomy-of-an-item/#attributes) 

Rollback:

* Time 5. Auth servers v1 read and write resources v0, migrate resources to v1 to the same path. Migration crashed.
* Time 6. Rollback, traffic sent back to Auth servers v0. Auth servers v0 read and write to attributes v0

Success:

* Time 5. Auth servers v1 read and write resources v0, migrate resources to v1 to the same path. Migration succeeded.
* Time 6. Auth servers v0 are shut down, All attributes version v0 are deleted, migration marked as completed.

The disadvantage of this model is that it requires 2x memory, but it trades it for availability, which is OK, if users want more availability. 

It will also allow lazy migrations, if you instruct auth server to read from the last available present value until background routine adds new and deletes old attributes

Reply:
Only one auth server reading and writing during an upgrade doesn't sound much different than scaling down to 1 auth during the upgrade. Won't this suffer from the same drawbacks we have today? (Uneven load, single point of failure, etc)

Reply:
@zmb3 you can have N > 1 auth servers in this design, as long as they are all the same version, that's the only constraint I propose to add. I am not proposing the constraint of having N=1 auth server that can write.

Reply:
Can we just require that all running auth servers are on the same version? It's trivially done in a kube deployment with a strategy type of "Recreate", and it shouldn't take long for pods to be spun back up on the new version. Asynchronous migrations will still allow us to spin two or more new auths at the same time (which is the biggest pain point of auth upgrades, right now), and downgrade if we notice problems until the migration is marked as finalized in some way and the old data is cleaned up.

Reply:
We could even have some config option that tells the auth servers to start doing background migrations automatically at startup or if they should be commanded externally (so we can avoid thrashing all the tenants at once).

Reply:
We could choose not to provide zero-downtime for Teleport auth upgrades when migrations are present. I'm not sure if this RFD would be a priority in that case, since it would have no impact on downtime, and my understanding is that zero-downtime is the primary motivation for making this change in the near future. Allowing multiple instances of a new version to start at the same time would not reduce auth downtime.

Reply:
@sclevine good points regarding possible downtime. How about this modified version:

* Time 0. All auth servers v0 are running. The system redirects traffic to v0 servers only.
* Time 1. All auth servers v0 are running. All auth servers v1 are running. The system redirects traffic to v0 and v1, but v1 is reading and writing in v0 compatibility mode.
* Time 2. Some auth servers v0 are running. All auth servers v1 are running and receiving all new traffic. The system is draining off requests to v0.
* Time 3. The system has drained off all traffic auth servers v0. The auth servers v1 are switching to v1 read/write mode and running migrations in the background.

In this case there won't be a scenario when both v0 and v1 servers are running and reading/writing at the same time with different resource versions.

@rosstimothy the gradual per resource migration seems very complicated and quite fragile, if any customer makes a mistake, we would have really hard time recovering the full chain.

Reply:
For the dynamodb backend, AWS strongly discourages connecting more than two simultaneous streams[1], so surging 100% could cause performance issues during upgrades. That said, we could achieve something similar via RollingUpdate with maxUnavailable=50% and maxSurge=0%, and by ensuring that v1 auth servers don't flip to v1 mode until all v0 auth servers are terminated. Where `*` = v0 mode:
```
v0 v0 -> v0 -> v0 v1* -> v1* -> v1 -> v1 v1
```
This wouldn't help with the thundering herd problem on large deployments, but it should provide continuous availability.

That said, @fspmarshall was pushing back on ever running multiple versions of auth at the same time. I'm not sure if the proposed compatibility mode addresses all of the issues he's brought up.

[1] https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html

Reply:
> For the dynamodb backend, AWS strongly discourages connecting more than two simultaneous streams[1], so surging 100% could cause performance issues during upgrades.

Our stream polling happens at a rate of 1 req/s per Auth server - I suspect the true limit for Dynamo streams is 5 req/s and is unrelated to the count of unique consumers. There's also a limit on stream throughput of 2MB/s so we'd need to validate the ability to surge against a tenant with a decently sized main table if migration will result in lots of writes while we've got extra consumers.